### ISSUE TITLE
Nodemon Fails to Wait for Process Exit when using Docker + Inspector

### ISSUE THREAD

**Issue by darkobits** [remy/nodemon#1476]
> Created per @remy's request in [this issue](https://github.com/remy/nodemon/issues/1346#issuecomment-442367096).

It seems that when restarting a process, Nodemon isn't waiting for the old process to completely exit before starting a new one, but:

- _only_ when in a Docker container and
- _only_ when the inspector/debug mode is active.

This issue can be intermittent in applications without a shutdown handler. But, having an application that implements a listener for Nodemon's `SIGUSR2` signal that takes even a second or so to complete can be enough to cause this issue to happen reliably.

The presence of `--inspect` _does_ cause the new process to immediately crash because its trying to open a debug server on the same port that is still in use by the previous process' debug server, but this merely a side-effect of this issue.

If running Nodemon in a Docker container without `--inspect`, Nodemon correctly waits for the old process to exit, but when `--inspect` is used inside the container, Nodemon _does not_ wait for the process to exit, leading to an immediate crash.

This can be reproduced with Nodemon `1.18.7` and the following:

```
$ node -v
v10.13.0

$ docker -v
Docker version 18.09.0, build 4d60db4

$ docker-compose -v
docker-compose version 1.23.1, build b02f1306
```

Here is a repo that replicates this issue: https://github.com/darkobits/nodemon-restart-issue.

**Comment by ImTheDeveloper** [comment]
Just started to see this issue when I upgraded to latest versions of nodemon too.

**Comment by remy** [comment]
I think this _might_ be solved with the next release going up [at the moment](https://travis-ci.org/remy/nodemon/builds/466125111) - assuming it passes, it'll land in nodemon@1.18.8.

@darkobits can you test this once it's up?

**Comment by denic** [comment]
@remy Works for me. Thanks!

**Comment by ImTheDeveloper** [comment]
Great thank you for this fix!

**Comment by bighuggies** [comment]
I'm still seeing the issue with `1.18.8`

**Comment by daniele-orlando** [comment]
`v1.18.8`

Same here, less frequent but still there.

**Comment by darkobits** [comment]
I've updated the [demo repo](darkobits/nodemon-restart-issue) referenced in the original post to nodemon `1.18.9`, and this issue is still occurring.

**Comment by darkobits** [comment]
@remy, thoughts regarding leaving this closed vs. re-opening?

**Comment by justinpage** [comment]
I'm seeing intermittent results. Sometimes when I save a file change, it restarts correctly. Other times it errors out the following message: 

```
crumb_1     | Starting inspector on 0.0.0.0:9229 failed: address already in use
crumb_1     | [nodemon] app crashed - waiting for file changes before starting...
crumb_1     | [nodemon] restarting due to changes...
crumb_1     | [nodemon] starting `node --inspect=0.0.0.0:9229 ./index.js`
```

What I'm forced to do, as shown above, is save again which triggers nodemon to restart. The second time around works correctly.

- version: `1.18.9`
- command `./node_modules/.bin/nodemon --inspect=0.0.0.0:9229 ./index.js`

**Comment by justinpage** [comment]
Still applicable.

**Comment by Genovo** [comment]
```
lsof -i tcp:9229
kill -9 <PID#>
```

**Comment by daniele-orlando** [comment]
Still persista

**Comment by anodynos** [comment]
I still get this error with `nodemon@1.18.10` & latest `node@10`

**Comment by d-wagner** [comment]
I also have this problem with `nodemon@1.18.9` and `1.8.10`

**Comment by aaronup** [comment]
We are also having this exact issue with `nodemon@1.18.10`

**Comment by abdohossam5** [comment]
Still persists

**Comment by yjwong** [comment]
We're also having this issue with `nodemon@1.18.10`. 

Reading the `nodemon` code, it seems to be the case that nodemon *does* wait for the child process to exit. The new process is only started in the exit handler for the child process (see https://github.com/remy/nodemon/blob/master/lib/monitor/run.js#L202 - `restart` calls `run` again).

When trying to `exec('ps -ef', (err, stdout) => console.log(stdout))` after the `spawn` call [here](https://github.com/remy/nodemon/blob/master/lib/monitor/run.js#L112), it appears that the old child process is either in the Z (zombie)/X (dead) state when this happens

So far, I've hit a dead end trying to debug this problem, but hopefully it's of help to someone.

**Comment by nomcopter** [comment]
This happens to me completely outside of docker. Repro repo here:  https://github.com/nomcopter/nodemon-no-wait-repro

**Comment by nomcopter** [comment]
Issue persists

**Comment by alemarcha** [comment]
It's not working here either

**Comment by nomcopter** [comment]
Repro repo that fails outside of docker on OS X @remy https://github.com/nomcopter/nodemon-no-wait-repro

**Comment by nomcopter** [comment]
Issue persists

**Comment by remy** [comment]
@nomcopter I've (finally) added a config to get stalebot to back off completely. I'll check out your repo on replication today (sorry for the lack of attention!)

**Comment by ChristianIvicevic** [comment]
I am unable to replicate this issue as well. Using node 10.15.3, 12.1.0 with latest nodemon and OS X 10.14.4. Will try later on Windows because I have the same issue in a different setting on Windows.

My own application only listens to the signals on Windows when Ctrl+C-ing, but does not recognize the signals when nodemon restarts. On my Mac it's actually the other way round: restarts are recognized followed by proper cleanup but Ctrl+C just kills everything outright without any cleanup.

**Comment by ChristianIvicevic** [comment]
Quick update: instead of `process.once` I had to use `process.on` in my own app to track a SIGINT via Ctrl+C on macOS which is rather confusing. Will report on the Windows behaviour for the repro repo later this evening.

**Comment by ChristianIvicevic** [comment]
Alright, I just tried it on Windows and the repro repo does not work as intended. Nodemon just kills the process without any cleanup at all and I think this issue isn't actually related to Nodemon but Node and Windows in general

**Comment by mdobroggg** [comment]
I'm having the same issue. I think it has to do with using an async call in process.on to exit. It seems like nodemon is acting on the original kill signal rather than the one re-thrown from process.on

**Comment by nomcopter** [comment]
@remy welp. Will try again locally when I have a change, thanks for taking a look.

**Comment by mdobroggg** [comment]
I think adding a small delay between kill and restart would be enough to fix this issue. If I cause nodemon to restart once I will get the error, but if I cause a restart after that failure, it will succeed

**Comment by pepe79** [comment]
Hello, today I worked on this and I think I have understood what is happening here.

First of all it seems, that this is only a Linux related issue (from the comments I see docker and Ubuntu). I had this problem since several weeks with Ubuntu 18. 

The problem is, that even if you think nodemon starts only one child process, nodemon actually starts two processes: one sh-process AND one node-process (run.js:97). The child.on('exit')-callback in run.js only gets events for the sh-Process (because it's the parent of the node process).
The node documentation explicitly says, that on Linux platform terminating the shell command does not necessarily kill all the shell child commands: 

> On Linux, child processes of child processes will not be terminated when attempting to kill their parent. This is likely to happen when running a new process in a shell or with the use of the shell option of ChildProcess.

(see https://nodejs.org/api/child_process.html#child_process_subprocess_kill_signal)
Although here the native kill command is executed in line run.js:368, it seems it behaves the same way.

Explicitly waiting for all other `kids` to be terminated before restarting the application solves the problem.

See my PR #1579

**Comment by daniele-orlando** [comment]
@pepe79 Thank you very much for having worked on the issue and the extensive explanation (very interesting and I've learned something new)

**Comment by pavles6** [comment]
Still persits, running on docker:

`[nodemon] Internal watch failed: ENOSPC: System limit for number of file watchers reached, watch '/sys/kernel/slab/:A-0000040/cgroup/pde_opener(667:console-setup.service)/cpu_partial'`

**Comment by aymather** [comment]
I've been having this issue for a while and my workaround might provide some information.

Right before I run `npm run dev` I run the following command: `lsof -ti tcp:5000 | xargs kill` which basically kills whatever processes are on the port.

However, even when I manually kill the port, I sometimes still get the same :::5000 address in use error. In that case I just kill the port again and it will usually work on the second try.

I'm just confused why even when I manually kill the port, I still get the error, and makes me think that maybe nodemon is the one blocking the port?

I'm running node v8.11.3 on MacOS 10.14.3.

**Comment by Remzi1993** [comment]
> I'm unable to replicate with OS X - see screenshot.
> 
> @nomcopter I'm using
> 
> * node@v10.14.1
> * nodemon@latest
> * OS X @ 10.13.6 (High Sierra)
> 
> ![Screen Shot 2019-05-08 at 14 21 34](https://user-images.githubusercontent.com/13700/57378406-a1820380-719c-11e9-80a5-b1d58e7e0a51.png)

Because the issue is on Linux! I hope that delaying it could fix this problem. It seems like @pepe79 has fixed this issue, see his forked repo and commit: https://github.com/pepe79/nodemon/commit/f6c28417d6fe5a82208e9c03a7ebb373e7cb7365 (https://github.com/pepe79/nodemon)

@remy I hope you could take a look at his repo/commit.

**Comment by towerofnix** [comment]
now that #1632 (which contains 0e6ba3ce6d3ad19f1348e547d15136b3c789732b) has been merged, should this issue be closed?

**Comment by remy** [comment]
👍

**Comment by marcelkottmann** [comment]
Hello @remy , sorry to say that, but this issue is back again since version 2.0.2. It was fixed in 2.0.0 and 2.0.1. I just had the chance to upgrade to version 2.0.2 and this issue is back again. I looked at 47dfb8be3ff06184793a55d32db4b6171fa2993f and I guess the new line
`child.kill(signal);`
should only be called after all sub-processes have already been terminated. Moving this line behind the waitForSubProcesses-function call, should fix this.
Otherwise the exit signal of child will restart the process tooo early (again).
